proxy/src/main/java/com/wavefront/agent/core/buffers/ActiveMQBuffer.java

```diff
@@ -0,0 +1,401 @@
+        new JmxGauge(addressObjectName, "AddressLimitPercent"));
+
+    countMetrics.put(
+        queue.getName(),
+        (PointsGauge)
+            Metrics.newGauge(
+                new MetricName("buffer." + name + "." + queue.getName(), "", "points"),
+                new PointsGauge(queue, activeMQServer)));
+
+    timeMetrics.put(
+        queue.getName(),
+        Metrics.newHistogram(
+            new MetricName("buffer." + name + "." + queue.getName(), "", "queue-time")));
+  }
+
+  public void shutdown() {
+    try {
+      for (Map.Entry<String, Session> entry : producers.entrySet()) {
+        entry.getValue().close(); // session
+        entry.getValue().close(); // producer
+      }
+      for (Map.Entry<String, Session> entry : consumers.entrySet()) {
+        entry.getValue().close(); // session
+        entry.getValue().close(); // consumer
+      }
+
+      activeMQServer.stop();
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void sendPoints(String queue, List<String> points) throws ActiveMQAddressFullException {
+    try {
+      doSendPoints(queue, points);
+    } catch (ActiveMQAddressFullException e) {
+      slowLog.log(Level.SEVERE, "Memory Queue full");
+      if (slowLog.isLoggable(Level.FINER)) {
+        slowLog.log(Level.SEVERE, "", e);
+      }
+      if (nextBuffer != null) {
+        nextBuffer.sendPoints(queue, points);
+        QueueStats.get(queue).queuedFull.inc();
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  public void doSendPoints(String queue, List<String> points) throws ActiveMQAddressFullException {
+    String sessionKey = "sendMsg." + queue + "." + Thread.currentThread().getName();
+    Session mqCtx =
+        producers.computeIfAbsent(
+            sessionKey,
+            s -> {
+              try {
+                ServerLocator serverLocator =
+                    ActiveMQClient.createServerLocator("vm://" + serverID);
+                ClientSessionFactory factory = serverLocator.createSessionFactory();
+                ClientSession session = factory.createSession();
+                ClientProducer producer = session.createProducer(queue);
+                return new Session(session, producer, serverLocator);
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+
+    try {
+      ClientMessage message = mqCtx.session.createMessage(true);
+      message.writeBodyBufferString(String.join("\n", points));
+      message.putIntProperty("points", points.size());
+      mqCtx.producer.send(message);
+    } catch (ActiveMQAddressFullException e) {
+      log.log(Level.FINE, "queue full: " + e.getMessage());
+      throw e;
+    } catch (ActiveMQObjectClosedException e) {
+      log.log(Level.FINE, "connection close: " + e.getMessage());
+      mqCtx.close();
+      producers.remove(sessionKey);
+      sendPoints(queue, points);
```

Comment:
Why do we call `sendPoints` here, when we call `doSendPoints` on line 184?  Should this be `nextBuffer.sendPoints` (line 190-193)?

Reply:
the call to `sendPoints` to retry after connection problem to activeMQ, but yes, if there is nextBuffer is safe to send it there.
Anyway, is a very very edge case.